Frontend/HsRenamer.hs

```diff
@@ -200,14 +200,26 @@ rnTerm (TmLet x tm1 tm2)  = do
   return (TmLet rnx rntm1 rntm2)
 rnTerm (TmCase scr alts)  = TmCase <$> rnTerm scr <*> mapM rnAlt alts
 
+-- | Rename a pattern
+-- | Turn this into tuple and check the "allDistint" here
+rnPat :: PsPat -> RnM (RnPat, [(PsTmVar, RnTmVar)])
+rnPat (HsPatCons dc ps) = do
+  rndc       <- lookupDataCon dc
+  psBindList <- mapM rnPat ps
+  let binds = concat $ map snd psBindList
+  if not $ distinct $ map fst binds
+    then throwErrorRnM (text "Term variables are not distict")
+    else return (HsPatCons rndc (map fst psBindList), binds)
+rnPat (HsPatVar x) = do
+  rnX <- rnTmVar x
+  return (HsPatVar rnX, [(x, rnX)])
+
```

Comment:
Instead of the hassle of `map fst` and `map snd` here and there I think it's just better to replace your
```... <- mapM rnPat ps```
with
```... <- mapAndUnzipM rnPat ps```
More about it [here](https://hackage.haskell.org/package/base-4.12.0.0/docs/Control-Monad.html#v:mapAndUnzipM).